### PR TITLE
Document new Datalog solver

### PR DESCRIPTION
**Pull request by bdahse**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Looks interesting. Looking forward to reading it carefully, hopefully soon.

**Comment by bdahse** [comment]
> 
> 
> Looks interesting. Looking forward to reading it carefully, hopefully soon.

It's not complete but the parts I have commented are (modulo review / questions).

**Comment by magnus-madsen** [comment]
Some tests are now broken?

**Comment by bdahse** [comment]
> 
> 
> Some tests are now broken?

Yeah, I broke it. Thats what I get for mixing in code changes :D

**Comment by bdahse** [comment]
Okay I get what's going on now. I didn't actually break it, my change fixed the bug I intended... and that exposed an old bug!

**Comment by magnus-madsen** [comment]
Will you fix the bug and then I can merge?

File: main/src/library/Fixpoint/Compiler.flix [hunk 1]
```diff
@@ -34,8 +42,24 @@ namespace Fixpoint {
...
+    ///     until(⋃Pi == ∅) do
+    ///     2.1. For each rule Ri with Pi in its head:
+    ///         purge ΔPi';
+    ///         ΔPi' = ΔPi' ∪ Eval-Rule-Incr(Ri);
+    ///         merge ΔPi' into Pi;
+    ///         ΔPi = ΔPi';
+    ///     end
+    /// Note that Eval-Rule is the code emitted by `compileRule`
+    /// and Eval-Rule-Incr is the code emitted by `compileRuleIncr`.
+    ///
```
Comment (magnus-madsen) [inline main/src/library/Fixpoint/Compiler.flix:62]: Very nice. I like the pseudo code outline.

**Review by magnus-madsen — COMMENTED** [review]
I added a few comments.

Overall I find that the quality of the writing is very good and appreciate the small examples throughout :)

**Comment by bdahse** [comment]
> 
> 
> Will you fix the bug and then I can merge?

I'm working on a fix, it's just a nasty bug that breaks one of my assumptions.

**Comment by bdahse** [comment]
I wanted to get the fix in even if I'm not happy with the code (why I didn't bother to write docs for the new code). It works but I welcome any questions or suggestions for improvement.

File: main/src/library/Fixpoint/Compiler.flix [hunk 2]
```diff
@@ -185,11 +261,22 @@ namespace Fixpoint {
...
     /// Guard atoms are compiled to function calls.
+    /// For example, a body like `not A(x), B(x), if x > 0.` is compiled to the following expressions:
+    /// 1.    B$1[0] = B$1[0]
+    /// 2.    (B$1[0]) ∉ A
+    /// 3.    <clo>(x)
+    /// Intuitively, the variable `x` has been unified with the term B$1[0]
+    /// because `x` occurs positively in the second atom.
+    /// All positive occurences of `x` is therefore equated with B$1[0], generating (1).
+    /// (2) comes from the negative atom `not A(x)`.
+    /// (3) is a function call that computes the expression `x > 0`.
```
[the remark this answers is not in the document]
Comment (bdahse) [reply main/src/library/Fixpoint/Compiler.flix:278]: The text after the point list is supposed to be an explanation of the point list's contents.

I think I understand your confusion. Some of the explanation relies on code outside the scope of this function (such as the unification of `x` with the term `B$1[0]`. That is handled by `unifyVars`).

Or does the confusion arise from a poorly chosen example?

File: main/src/library/Fixpoint/IndexSelection.flix [hunk 3]
```diff
@@ -16,36 +16,41 @@
...
+    ///                     if (z = u) then
+    ///                         project x into a
+    /// After step 2:
+    /// search x ∈ b do
+    ///     if (x ∉ a) then
+    ///         query {(y, z, w) ∈ r | x = y} do
+    ///             if (x = w) then
+    ///                 query {u ∈ c | z = u} do
+    ///                     project x into a
+    ///
```
[the remark this answers is not in the document]
Comment (bdahse) [reply main/src/library/Fixpoint/IndexSelection.flix:51]: Yes. Should I adopt a convention where `A` names a relation and `a` names a variable?

File: main/src/library/Fixpoint/Solver.flix [hunk 4]
```diff
@@ -37,27 +47,19 @@ namespace Fixpoint {
          let _ = notifyPostSolve(model);
          model
 
+    ///
+    /// Return the pairwise union of `d1` and `d2`.
+    /// I.e. the facts of the union is the union of the facts and likewise for rules.
+    /// Facts and rules are not deduplicated. This has no effect on the minimal model.
```
[the remark this answers is not in the document]
Comment (bdahse) [reply main/src/library/Fixpoint/Solver.flix:53]: Is this function even relevant anymore?

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Fixpoint/Solver.flix]: I think so, when you do `solve e1, e2` or `query e1, e2` and so forth?

**Comment by magnus-madsen** [comment]
Ready for review/merge?

**Comment by bdahse** [comment]
> 
> 
> Ready for review/merge?

Yes.

File: main/src/library/Fixpoint/IndexSelection.flix (hunk 3 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Fixpoint/IndexSelection.flix:58]: Yes, that would be useful. I see you already did it!

**Comment by magnus-madsen** [comment]
Merged. Thanks!

**Pull request merged**